pkg/controllermanager/controller/controllerregistration/controllerregistration_seed_control.go

```diff
@@ -0,0 +1,458 @@
+// to be installed. The computation is performed based on a list of required kind/type combinations and the proper mapping
+// to existing ControllerRegistration objects.
+func computeWantedControllerRegistrationNames(
+	wantedKindTypeCombinations sets.String,
+	kindTypeToControllerRegistrationName map[string]string,
+) (sets.String, error) {
+	names := sets.NewString()
+
+	for _, requiredExtension := range wantedKindTypeCombinations.UnsortedList() {
+		name, ok := kindTypeToControllerRegistrationName[requiredExtension]
+		if !ok {
+			return nil, fmt.Errorf("need to install an extension controller for %q but no appropriate ControllerRegistration found", requiredExtension)
+		}
+
+		names.Insert(name)
+	}
+
+	return names, nil
+}
+
+// computeRegistrationNameToInstallationNameMap computes a map that maps the name of a ControllerRegistration to the name of an
+// existing ControllerInstallation object that references this registration.
+func computeRegistrationNameToInstallationNameMap(
+	controllerInstallationList *gardencorev1beta1.ControllerInstallationList,
+	controllerRegistrationNameToObject map[string]*gardencorev1beta1.ControllerRegistration,
+	seedName string,
+) (map[string]string, error) {
+	registrationNameToInstallationName := make(map[string]string)
+
+	for _, controllerInstallation := range controllerInstallationList.Items {
+		if controllerInstallation.Spec.SeedRef.Name != seedName {
+			continue
+		}
+
+		if _, ok := controllerRegistrationNameToObject[controllerInstallation.Spec.RegistrationRef.Name]; !ok {
+			return nil, fmt.Errorf("ControllerRegistration %q does not exist", controllerInstallation.Spec.RegistrationRef.Name)
+		}
+
+		registrationNameToInstallationName[controllerInstallation.Spec.RegistrationRef.Name] = controllerInstallation.Name
+	}
+
+	return registrationNameToInstallationName, nil
+}
+
+// deployNeededInstallations takes the list of required names of ControllerRegistrations, a mapping of ControllerRegistration
+// names to their actual objects, and another mapping of ControllerRegistration names to existing ControllerInstallations. It
+// creates or update ControllerInstallation objects for that reference the given seed and the various desired ControllerRegistrations.
+func deployNeededInstallations(
+	ctx context.Context,
+	logger *logrus.Entry,
+	c client.Client,
+	seed *gardencorev1beta1.Seed,
+	wantedControllerRegistrations sets.String,
+	controllerRegistrationNameToObject map[string]*gardencorev1beta1.ControllerRegistration,
+	registrationNameToInstallationName map[string]string,
+) error {
+	for _, registrationName := range wantedControllerRegistrations.UnsortedList() {
+		logger.Infof("Deploying wanted ControllerInstallation for %q", registrationName)
+
+		if err := deployNeededInstallation(ctx, c, seed, registrationName, controllerRegistrationNameToObject, registrationNameToInstallationName); err != nil {
```

Comment:
Why do you pass down the whole `controllerRegistrationNameToObject` map? Can't you simply pass down the needed registration object? Same for the installation map...

```suggestion
		if err := deployNeededInstallation(ctx, c, seed, registrationName, controllerRegistrationNameToObject[registrationName], registrationNameToInstallationName[registrationName]); err != nil {
```